### ISSUE TITLE
TS4.1.2 Regression: Incorrect arity inferred on function returned by util.callbackify if input function has optional or default argument

### ISSUE THREAD

**Issue by whoisbenli** [microsoft/TypeScript#41607]
**TypeScript Version:**  v4.2.0-dev.20201119. Also impacts: 4.1.2. Does not impact: 4.1.0-beta, 4.0.5

**Search Terms:**
callbackify expected arguments optional default

**Code**

```ts
import util from 'util';

function functionWithDefaultArgReturnsPromise(arg1: number, arg2: string, arg3: boolean = false): Promise<number> {
   return new Promise<number>((resolve, reject) => {
      // do something
   });
}

const functionWithDefaultArgHasCallBack = util.callbackify(functionWithDefaultArgReturnsPromise)

functionWithDefaultArgHasCallBack(1, 'two', true, (err: any, result: any) => {
   // do something

   // this callback function triggers error 
   //
   //    "Expected 3 arguments, but got 4.(2554)"
   // 
   // in the v4.2.0-dev.20201119 (nightly at the time in the playground),
   // but not with 4.1.0-beta in the playground, but I do see it on 4.1.2 locally as well as the last release candidate.
});
```

**Expected behavior:**
No error, no red squiggly lines

**Actual behavior:**
Error "Expected 3 arguments, but got 4.(2554)" occurs with red squiggly underlines on the callbackified function.

**Playground Link:** 
4.2.0-dev.20201119 - Error seen
https://www.typescriptlang.org/play?ts=4.2.0-dev.20201119#code/JYWwDg9gTgLgBAVxsANnAZlCI4HImq4DcAUCeggHYDGyElGVtw9A6sDABYAiApugEMEKGAEEoAcwBKvGAiiUAzgAUsIYIt4AKAZICMALjiUEIAEa8oAGji6JAJiOKYUYJQk27AZiNmIEFF4BBgBeDAEUTQBKI1VsDV4AHhNzSwA+OABvEjhcqFl5BkpeAHc4OPVNZNMLKDStLXzFAIA3Xht8gCteWii4EIzs3OG4AHpRuAATCDhmkFk3CRzcgF8o0hWyanpnRho6SnYuPkFhMUkACQFFAGEIlAAhAWoAa37EZBQAOmp7s2eXsB0ABPLQUfYsQ4cHj8IQicTSAoKFRqBJRMjg5hsaEnOHnCRXW73J6vLR6Gy4GAlCC4GwuBDtOBaSzWOBNM59AZZZZjCbTWbYWScRZkYbjOBcDRwX4oFD-V57LEMFzACQSSyKOA88baibDABEAFEAB5gHowXiTOBeWySUy8SgwRQ2MxIOASCDwAAsXy09gArP6vVF9bqtWKJm4JZxeHAWj77F8AAwAWkmvBaX3sSezejzAE4mZRVZwYChgbZ4FxY8h5lErGHXfBKJ64CVoXAfXpkymLDABC63QBJKYzTSxjhweidr7d+xwFAQGXlr4kNZEIA

4.1.2 - Error seen
https://www.typescriptlang.org/play?ts=4.1.2#code/JYWwDg9gTgLgBAVxsANnAZlCI4HImq4DcAUCeggHYDGyElGVtw9A6sDABYAiApugEMEKGAEEoAcwBKvGAiiUAzgAUsIYIt4AKAZICMALjiUEIAEa8oAGji6JAJiOKYUYJQk27AZiNmIEFF4BBgBeDAEUTQBKI1VsDV4AHhNzSwA+OABvEjhcqFl5BkpeAHc4OPVNZNMLKDStLXzFAIA3Xht8gCteWii4EIzs3OG4AHpRuAATCDhmkFk3CRzcgF8o0hWyanpnRho6SnYuPkFhMUkACQFFAGEIlAAhAWoAa37EZBQAOmp7s2eXsB0ABPLQUfYsQ4cHj8IQicTSAoKFRqBJRMjg5hsaEnOHnCRXW73J6vLR6Gy4GAlCC4GwuBDtOBaSzWOBNM59AZZZZjCbTWbYWScRZkYbjOBcDRwX4oFD-V57LEMFzACQSSyKOA88baibDABEAFEAB5gHowXiTOBeWySUy8SgwRQ2MxIOASCDwAAsXy09gArP6vVF9bqtWKJm4JZxeHAWj77F8AAwAWkmvBaX3sSezejzAE4mZRVZwYChgbZ4FxY8h5lErGHXfBKJ64CVoXAfXpkymLDABC63QBJKYzTSxjhweidr7d+xwFAQGXlr4kNZEIA

4.0.5 - Error not seen
https://www.typescriptlang.org/play?ts=4.0.5#code/JYWwDg9gTgLgBAVxsANnAZlCI4HImq4DcAUCeggHYDGyElGVtw9A6sDABYAiApugEMEKGAEEoAcwBKvGAiiUAzgAUsIYIt4AKAZICMALjiUEIAEa8oAGji6JAJiOKYUYJQk27AZiNmIEFF4BBgBeDAEUTQBKI1VsDV4AHhNzSwA+OABvEjhcqFl5BkpeAHc4OPVNZNMLKDStLXzFAIA3Xht8gCteWii4EIzs3OG4AHpRuAATCDhmkFk3CRzcgF8o0hWyanpnRho6SnYuPkFhMUkACQFFAGEIlAAhAWoAa37EZBQAOmp7s2eXsB0ABPLQUfYsQ4cHj8IQicTSAoKFRqBJRMjg5hsaEnOHnCRXW73J6vLR6Gy4GAlCC4GwuBDtOBaSzWOBNM59AZZZZjCbTWbYWScRZkYbjOBcDRwX4oFD-V57LEMFzACQSSyKOA88baibDABEAFEAB5gHowXiTOBeWySUy8SgwRQ2MxIOASCDwAAsXy09gArP6vVF9bqtWKJm4JZxeHAWj77F8AAwAWkmvBaX3sSezejzAE4mZRVZwYChgbZ4FxY8h5lErGHXfBKJ64CVoXAfXpkymLDABC63QBJKYzTSxjhweidr7d+xwFAQGXlr4kNZEIA

**Related Issues:**
No related issue found.

**Comment by orta** [comment]
Here's a compiler backed repro:

````ts repro
function functionWithDefaultArgReturnsPromise(arg1: number, arg2: string, arg3: boolean = false): Promise<number> {
    return new Promise<number>((resolve, reject) => {
        // do something
    });
}

const functionWithDefaultArgHasCallBack = callbackify(functionWithDefaultArgReturnsPromise)

functionWithDefaultArgHasCallBack(1, 'two', true, (err, result) => {
    // do something

    // this callback function triggers error 
    //
    //    "Expected 3 arguments, but got 4.(2554)"
    // 
    // in the v4.2.0-dev.20201119 (nightly at the time in the playground),
    // but not with 4.1.0-beta in the playground, but I do see it on 4.1.2 locally as well as the last release candidate.
});

// C&P'd infra from node's types below

interface ErrnoException extends Error {
    errno?: number;
    code?: string;
    path?: string;
    syscall?: string;
    stack?: string;
}

function callbackify(fn: () => Promise<void>): (callback: (err: ErrnoException) => void) => void;
function callbackify<TResult>(fn: () => Promise<TResult>): (callback: (err: ErrnoException, result: TResult) => void) => void;
function callbackify<T1>(fn: (arg1: T1) => Promise<void>): (arg1: T1, callback: (err: ErrnoException) => void) => void;
function callbackify<T1, TResult>(fn: (arg1: T1) => Promise<TResult>): (arg1: T1, callback: (err: ErrnoException, result: TResult) => void) => void;
function callbackify<T1, T2>(fn: (arg1: T1, arg2: T2) => Promise<void>): (arg1: T1, arg2: T2, callback: (err: ErrnoException) => void) => void;
function callbackify<T1, T2, TResult>(fn: (arg1: T1, arg2: T2) => Promise<TResult>): (arg1: T1, arg2: T2, callback: (err: ErrnoException | null, result: TResult) => void) => void;
function callbackify<T1, T2, T3>(fn: (arg1: T1, arg2: T2, arg3: T3) => Promise<void>): (arg1: T1, arg2: T2, arg3: T3, callback: (err: ErrnoException) => void) => void;
function callbackify<T1, T2, T3, TResult>(
    fn: (arg1: T1, arg2: T2, arg3: T3) => Promise<TResult>): (arg1: T1, arg2: T2, arg3: T3, callback: (err: ErrnoException | null, result: TResult) => void) => void;
function callbackify<T1, T2, T3, T4>(
    fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4) => Promise<void>): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, callback: (err: ErrnoException) => void) => void;
function callbackify<T1, T2, T3, T4, TResult>(
    fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4) => Promise<TResult>): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, callback: (err: ErrnoException | null, result: TResult) => void) => void;
function callbackify<T1, T2, T3, T4, T5>(
    fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5) => Promise<void>): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, callback: (err: ErrnoException) => void) => void;
function callbackify<T1, T2, T3, T4, T5, TResult>(
    fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5) => Promise<TResult>,
): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, callback: (err: ErrnoException | null, result: TResult) => void) => void;
// @ts-ignore
function callbackify<T1, T2, T3, T4, T5, T6>(
    fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, arg6: T6) => Promise<void>,
): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, arg6: T6, callback: (err: ErrnoException) => void) => void;
function callbackify<T1, T2, T3, T4, T5, T6, TResult>(
    fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, arg6: T6) => Promise<TResult>
): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, arg6: T6, callback: (err: ErrnoException | null, result: TResult) => void) => void {
    throw new Error("")
}

```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?#code/GYVwdgxgLglg9mABKSsEHUZQBYBECmwAhiADZQCCATgOYBK+UIVYAzgApVwC2Mr+ACiK0AjAC5EYENwBG+KgBpEwmgCYJrKFRhgaSlQGYJMuHFL4iSALzIipfgEoJnHn3wAeKbPkA+RAG8AKEQQxCpGZiQwfAB3RBdefk9pOSofAQFw1jMAN3wlcIArfGgHRCs-INDqxAB6WsQAEzhEbO5GbB0aYNCAXwcAbkDewMCIBE1kcGh4MEwcAmIySloACSJWAGE7UgAhIggAa3LECB2ZA8OYYABPARQZjCw8QhJyanoIlg4uRPwHUYPNBzZ6LN4rGjrLY7fZHAQiJQAcigMTgiKUWhA+UQAnkijC+FYyzKFQCPRC9SaLTaHS6o2qlJwfFO50uU1Qs0QWhgNBo8lYiDxcCoiHJdVqYspoQARABRAAeAAcSlB8I1EAZlLRpPgwFBWEoZCAoIgaHATQAWAB0AlUAFY7RaHNLJQ1XYgdFzsPhEDlraorQAGAC0jXwOStqkDUZEsYAnDiwDzsFBSDdlCacD7YO0PUgs4hFaQiDcaFxwI0HAp3UaTWBzYgYs9ENaREHg3IoEQ816fUWS2W4BXDcbEABJKmtfA+rCIBAtq1t1SIUhwM6kNPKAUxfAbre9lcbE3hcwbH1nMCNGCNIiqq3DQajSmbABk7ER6p0wCo3e-PEkcBhoiApQDcyoCnIq4xKMOiqlQxAQD6spUCwcAKohirAoK8qqpeArIVwIpVKEeL1gA-BIXipEM1TjGGFGtNyug0aEiq3tgDGaNozFiqwNysOupCcUxNAsSEmiXMJ3GicMgLTFhgkXEc1x3MAYASAIJJ+AkbjuDkcDXj4Tg4oplwaXiEgEfW6H4Jhsxab6BmVuUfj6deQxApypnKbc7gACoMES5DpGpGkOTpSQBYSyxGRp3mHOZKGWSh1nyhhwIFNF5ASFFQVQA5bnOaShUefJXmsj5Nz+SIIXqTiKjiIgfkiOFvy6YVsX1aIOUIiyG5KQlOIWYgVloWltnAgVTlTe5gSefO8UqdVSi5TF9x1UI3VNS1LnxG1kWBTFxmbTQjXNUo8WJVQyWoTZdkIJleU5Yd5AzUVrlOaVHILRVVy+edTWqLVGkNT1+i0OogOta4SQdcdoPbeDag5aoF2-VdN2pel9m7YVb1fY8SCLf9vV+ajTUvVAwNdadYNasjUO7RFHircF8NbQDKiQ2TaP9WZQ1JSNKVjdj84AD6SGQpCPcsz1ZfluPTYrs3zUTv1LQDPNNQY1MnWdvVcyjSNGNr0N-HpTmdXrdOG4Dxs5QYvOkANGNC7d433WAb343NZU-XzlXLXb2srZT6RiqFNP60j3Pk4YDtm7prNU+ztOI-Tsf2yHfXO-zuKC6Nd1YRLUgbjL2UU-L3vK40BMKerJMreTfmO01Frh9UkfW+ntta-H2cqBaOVOkz+0eHDIMcwbENG-TJst0jQ9t07LsC9dbtYxNOPFUrO8q37asB39VWa83rd+Raofyx3oRdwjnMz8H-cL-TS8X4nB3X6n0cZ7Pz+t4PYeK887DULh7Yuksy4EiepXPK1c96119t9Q+udA6nxWufS+TU7Q3xCHfKeMc-60HngA2gb8sEqDtDlO0H9x6W2-jbR+fdiEO0XkA+mVDsHAKOK7MBosvY1x9qrHOA0NakzPitLBfk7RXzyrg5AG177TwZswmgJC2HLw4dQ2h-kw7VgYT3JhccWEDzIewyh1DuGDXzuvPhW9xaQOltA2WsDiSCJrkMSkAABfUwYeT1nCEgwmIjLhiKbhgyRK0ZFNQAGzyPwWnB+KjjFqNYa-cxtBOHSKRjEnKMSdEdX0ZPRJyjM5zzSYAzRFiuH01ybEqxvDhZF23h9a8QiD4hLQeIiJmjsn1NccFAQEdFEEN-k-ExL9KkXyRlk6JKg6l+XyaPGGLMw6BAMUksp-8NHTK0TU+ZeSGlr0xiLexSAS5S3LlAOWcD3EILJNUHAXA4jRDiARYUAhpTSgBCMIAA)

**Comment by ahejlsberg** [comment]
This is an effect of #41308. With that PR we made our handling of optional parameters in overload ladders more consistent. For example:

```ts
declare function foo(cb: () => void): 0;
declare function foo(cb: (x: number) => void): 1;
declare function foo(cb: (x: number, y: number) => void): 2;

declare function f1(x: number): void;
declare function f2(x: number, y?: number): void;
declare function f3(x: number, y?: any): void;
declare function f4(x: number, y?: unknown): void;

let z1 = foo(f1);  // 1
let z2 = foo(f2);  // was 2 (pre-4.1), now 1
let z3 = foo(f3);  // 1
let z4 = foo(f4);  // 1
```

It has to do with some pretty strange behavior in the subtype relation which we use in the first pass of overload resolution. Previously we didn't consider `(x: number, y?: number) => void` to be a subtype of `(x: number) => void`, but strangely we did consider `(x: number, y?: any) => void` and `(x: number, y?: unknown) => void` to be a subtypes. We now consider all of them subtypes. It's definitely more consistent, but it ends up subtly changing which overloads we pick in overload ladders when the source has optional parameters.

To put it in context of the OP example above, pre-4.1 we report an error when the function signature is changed to

```ts
function functionWithDefaultArgReturnsPromise(arg1: number, arg2: string, arg3: unknown = false): Promise<number>;
```

i.e. if type `unknown` is used instead of `boolean` for the optional argument. That seems very inconsistent. In 4.1 we consistently pick the first overload that matches, which means we consider the optional argument to be omitted.
